### PR TITLE
CPM-78: Add filter empty, not empty on product grid

### PR DESCRIPTION
**Pull request by pierallard**

### CODE REVIEW

**Review by SamirBoulil — APPROVED** [review]
Missing one integration test here tests/back/Pim/Enrichment/Integration/PQB/Filter/Attribute/Boolean/BooleanFilterIntegration.php

File: src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Filter/Attribute/BooleanFilter.php [hunk 1]
```diff
@@ -84,6 +84,15 @@ public function addAttributeFilter(
...
                 break;
 
+            case Operators::IS_EMPTY:
+                $mustNotClause = [
+                    'exists' => [
+                        'field' => $attributePath,
+                    ],
+                ];
+                $this->searchQueryBuilder->addMustNot($mustNotClause);
+                break;
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Filter/Attribute/BooleanFilter.php:94]: I think we should also check if the attribute belongs to the family. Like this for example: https://github.com/akeneo/pim-community-dev/blob/master/src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Filter/Attribute/NumberFilter.php#L119-L138

File: src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Filter/Attribute/BooleanFilter.php [hunk 2]
```diff
@@ -84,6 +84,15 @@ public function addAttributeFilter(
                 $this->searchQueryBuilder->addFilter($filterClause);
                 break;
 
+            case Operators::IS_EMPTY:
+                $mustNotClause = [
+                    'exists' => [
+                        'field' => $attributePath,
+                    ],
+                ];
+                $this->searchQueryBuilder->addMustNot($mustNotClause);
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Filter/Attribute/BooleanFilter.php:93]: here you will also get products whose family doesn't contain the attribute,
you can add a ` should`  clause  like 
```php
$attributeInEntityClauses = [
                    [
                        'terms' => [
                            self::ATTRIBUTES_FOR_THIS_LEVEL_ES_ID => [$attribute->getCode()],
                        ],
                    ],
                    [
                        'terms' => [
                            self::ATTRIBUTES_OF_ANCESTORS_ES_ID => [$attribute->getCode()],
                        ],
                    ]
                ];
                $this->searchQueryBuilder->addFilter(
                    [
                        'bool' => [
                            'should' => $attributeInEntityClauses,
                            'minimum_should_match' => 1,
                        ],
                    ]
                );

File: src/Oro/Bundle/PimFilterBundle/Form/Type/Filter/BooleanFilterType.php [hunk 3]
```diff
@@ -18,6 +18,8 @@ class BooleanFilterType extends AbstractChoiceType
     /** @staticvar integer */
     const TYPE_YES = 1;
     const TYPE_NO = 0;
+    const TYPE_EMPTY = 2;
+    const TYPE_NOT_EMPTY = 3;
```
Comment (nmarniesse) [inline src/Oro/Bundle/PimFilterBundle/Form/Type/Filter/BooleanFilterType.php:22]: The possible values and the operators are mixed. If it's not confusing I'm ok with that (I don't know)

**Review by nmarniesse — COMMENTED** [review]
Could you add a test in https://github.com/akeneo/pim-community-dev/blob/master/tests/back/Pim/Enrichment/Integration/PQB/Filter/Attribute/Boolean/BooleanFilterIntegration.php pls?  
Here the integration test is more important than the spec IMHO.

File: tests/back/Pim/Enrichment/Integration/PQB/Filter/Attribute/Boolean/BooleanFilterIntegration.php [hunk 4]
```diff
@@ -63,7 +63,7 @@ public function testOperatorNotEmpty()
     public function testOperatorEmpty()
     {
         $result = $this->executeFilter([['a_yes_no', Operators::IS_EMPTY, '']]);
-        $this->assert($result, ['empty']);
+        $this->assert($result, []); // The 'empty' product does not belong to a family having boolean as attribute.
```
Comment (mmetayer) [inline tests/back/Pim/Enrichment/Integration/PQB/Filter/Attribute/Boolean/BooleanFilterIntegration.php:66]: FYI, a better solution here IMO: https://github.com/akeneo/pim-community-dev/pull/13282/files#diff-d13bea837daff0105feb8973ece05052a485b72a36e8c96b6e74e9ce48d2c109R36

**Review by mmetayer — APPROVED** [review]

**Pull request merged**